### PR TITLE
add new EPL + 2 multipole profiles

### PR DESCRIPTION
**Pull request by dangilman**
I implement an EPL profile with two multipole terms concentric with the EPL origin

### CODE REVIEW

**Comment by dangilman** [comment]
@sibirrer all tests and coverage tests pass

File: lenstronomy/LensModel/Profiles/epl_multipole_m3m4.py [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+import lenstronomy.Util.param_util as param_util
+from lenstronomy.LensModel.Profiles.base_profile import LensProfileBase
+from lenstronomy.LensModel.Profiles.epl import EPL
+from lenstronomy.LensModel.Profiles.multipole import Multipole
+
+__all__ = ["EPL_MULTIPOLE_M3M4"]
+
+
+class EPL_MULTIPOLE_M3M4(LensProfileBase):
+    """EPL (Elliptical Power Law) mass profile combined with two multiple terms of order
```
Comment (sibirrer) [inline lenstronomy/LensModel/Profiles/epl_multipole_m3m4.py:13]: 'multipole'

File: lenstronomy/LensModel/Profiles/epl_multipole_m3m4.py [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+    m=3 and m=4.
+
+    Reference to the implementation: https://ui.adsabs.harvard.edu/abs/2022A%26A...659A.127V/abstract
+
+    See also documentation of EPL_BOXYDIKSY CLASS, lenstronomy.LensModel.Profiles.epl and
+    lenstrnomy.LensModel.Profiles.multipole for details.
```
Comment (sibirrer) [inline lenstronomy/LensModel/Profiles/epl_multipole_m3m4.py:19]: lenstronomy typo

File: lenstronomy/LensModel/Profiles/epl_multipole_m3m4.py [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+
+    :param theta_E: Einstein radius
+    :param gamma: negative power-law slope of the 3D mass distributions
+    :param e1: eccentricity. For details, read lenstronomy.Util.param_util.phi_q2_ellipticity document.
+    :param e2: eccentricity. For details, read lenstronomy.Util.param_util.phi_q2_ellipticity document.
+    :param center_x: center of distortion
+    :param center_y: center of distortion
+    :param a3_a: Strength of the deviation from elliptical isodensity contours caused by the multipole term of order 3
```
Comment (sibirrer) [inline lenstronomy/LensModel/Profiles/epl_multipole_m3m4.py:27]: can you do more documentation of the multipoles or refer to the specific equations if there are elsewhere?

File: lenstronomy/LensModel/Profiles/epl_multipole_m3m4.py [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+        a3_a,
+        delta_phi_m3,
+        a4_a,
+        delta_phi_m4,
+        center_x=0,
+        center_y=0,
+    ):
+        """Computes the gravitational potential in units of theta_E^2 :param x:
+
+        x-coordinate in image plane :param y: y-coordinate in image plane :param
```
Comment (sibirrer) [inline lenstronomy/LensModel/Profiles/epl_multipole_m3m4.py:157]: some docstrings got messed up here

**Review by sibirrer — CHANGES_REQUESTED** [review]
Looks good. I found some minor typos, formatting issue and perhaps the documentation of the multipole parameters can be improved by adding an equation

**Comment by dangilman** [comment]
> Looks good. I found some minor typos, formatting issue and perhaps the documentation of the multipole parameters can be improved by adding an equation

I made the requested changes, some formatting issue is preventing it from passing? not sure what's going on

**Comment by sibirrer** [comment]
Perhaps you can run the formatter locally? (make sure you update your docformatter):
```
flake8 lenstronomy test
black .
docformatter -r ./* --black --in-place
```

**Comment by dangilman** [comment]
> Perhaps you can run the formatter locally? (make sure you update your docformatter):
> 
> ```
> flake8 lenstronomy test
> black .
> docformatter -r ./* --black --in-place
> ```

I pulled the automatic formatting changes done on the pull request back to my local machine, and they are causing the weird formatting issues. The format looks correct on my branch

File: lenstronomy/LensModel/Profiles/epl_multipole_m3m4.py [hunk 5]
```diff
@@ -10,27 +10,27 @@
...
     :param gamma: negative power-law slope of the 3D mass distributions
     :param e1: eccentricity. For details, read lenstronomy.Util.param_util.phi_q2_ellipticity document.
     :param e2: eccentricity. For details, read lenstronomy.Util.param_util.phi_q2_ellipticity document.
     :param center_x: center of distortion
     :param center_y: center of distortion
-    :param a3_a: Strength of the deviation from elliptical isodensity contours caused by the multipole term of order 3
-    translated into the multipole strength from the MULTIPOLE class through a rescaling by theta_E / sqrt(q). The rescaling
+    :param a3_a: Strength of the deviation from elliptical isodensity contours caused by the multipole term of order 3. For details regarding the multipole mass profile see documentation in the class lenstronomy/LensModel/Profiles/multipole.py.
+    The multipole amplitude set by a3_a is translated into the multipole strength from the MULTIPOLE class through a rescaling by theta_E / sqrt(q). This rescaling
     preserves the shape of the isodensity contours such that a3_a produces the same shape regardless of theta_E or q.
```
Comment (ajshajib) [inline lenstronomy/LensModel/Profiles/epl_multipole_m3m4.py]: @dangilman, looks like you have put new lines in between lines of parameter definitions. This is most likely throwing the `docformatter` off.

Comment (ajshajib) [reply lenstronomy/LensModel/Profiles/epl_multipole_m3m4.py]: I think, if you reformat this docstring by hand without the new lines as before, `docformatter` would be able to format it, if still necessary, for the test to pass.

File: lenstronomy/LensModel/Profiles/epl_multipole_m3m4.py [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+    The multipole amplitude set by a3_a is translated into the multipole strength from the MULTIPOLE class through a rescaling by theta_E / sqrt(q). This rescaling
+    preserves the shape of the isodensity contours such that a3_a produces the same shape regardless of theta_E or q.
+    :param delta_phi_m3: angle of the m=3 multipole profile relative to the position angle of the EPL profile
+    :param a4_a: Strength of the deviation from elliptical isodensity contours caused by the multipole term of order 4. For details regarding the multipole mass profile see documentation in the class lenstronomy/LensModel/Profiles/multipole.py.
+    The multipole amplitude set by a4_a is translated into the multipole strength from the MULTIPOLE class through a rescaling by theta_E / sqrt(q).
+    This profile is disky when a4_a>0 and boxy when a4_a<0 for phi_m_a4a=0.0.
+    :param delta_phi_m4: angle of the m=4 multipole profile relative to the position angle of the EPL profile
```
Comment (ajshajib) [inline lenstronomy/LensModel/Profiles/epl_multipole_m3m4.py:34]:
```suggestion
    :param theta_E: Einstein radius
    :param gamma: negative power-law slope of the 3D mass distributions
    :param e1: eccentricity. For details, read lenstronomy.Util.param_util.phi_q2_ellipticity document.
    :param e2: eccentricity. For details, read lenstronomy.Util.param_util.phi_q2_ellipticity document.
    :param center_x: center of distortion
    :param center_y: center of distortion
    :param a3_a: Strength of the deviation from elliptical isodensity contours caused by the multipole term of order 3. For details regarding the multipole mass profile see documentation in the class lenstronomy/LensModel/Profiles/multipole.py. The multipole amplitude set by a3_a is translated into the multipole strength from the MULTIPOLE class through a rescaling by theta_E / sqrt(q). This rescaling preserves the shape of the isodensity contours such that a3_a produces the same shape regardless of theta_E or q.
    :param delta_phi_m3: angle of the m=3 multipole profile relative to the position angle of the EPL profile
    :param a4_a: Strength of the deviation from elliptical isodensity contours caused by the multipole term of order 4. For details regarding the multipole mass profile see documentation in the class lenstronomy/LensModel/Profiles/multipole.py. The multipole amplitude set by a4_a is translated into the multipole strength from the MULTIPOLE class through a rescaling by theta_E / sqrt(q). This profile is disky when a4_a>0 and boxy when a4_a<0 for phi_m_a4a=0.0.
    :param delta_phi_m4: angle of the m=4 multipole profile relative to the position angle of the EPL profile
```

File: lenstronomy/LensModel/Profiles/epl_multipole_m3m4.py [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+        profile (along 1st axis) :param e2: ellipticity of EPL profile (along 2nd axis)
+        :param a3_a: amplitude of the m=3 multiple deviation from pure elliptical shape
+        related to the physical amplitude of the MULTIPOLE profile by a scaling theta_E
+        / sqrt(q) :param delta_phi_m3: orientation of the m=3 profile relative to the
+        position angle of the EPL profile :param a4_a: amplitude of the m=4 multipole
+        deviation from pure elliptical shape related to the physical amplitude of the
+        MULTIPOLE profile by a scaling theta_E / sqrt(q) :param delta_phi_m4:
+        orientation of the m=4 profile relative to the position angle of the EPL profile
+        :param center_x: center of the profile :param center_y: center of the profile
+        :return: gravitational potential in units of theta_E^2."""
```
Comment (ajshajib) [inline lenstronomy/LensModel/Profiles/epl_multipole_m3m4.py:167]:
```suggestion
        """Computes the gravitational potential in units of theta_E^2 
        
        :param x: x angular coordinate 
        :param y: y angular coordinate 
        :param theta_E: Einstein radius 
        :param gamma: log-slope of EPL mass profile 
        :param e1: ellipticity of EPL profile (along 1st axis) 
        :param e2: ellipticity of EPL profile (along 2nd axis)
        :param a3_a: amplitude of the m=3 multiple deviation from pure elliptical shape related to the physical amplitude of the MULTIPOLE profile by a scaling theta_E / sqrt(q) 
        :param delta_phi_m3: orientation of the m=3 profile relative to the position angle of the EPL profile 
        :param a4_a: amplitude of the m=4 multipole deviation from pure elliptical shape related to the physical amplitude of the MULTIPOLE profile by a scaling theta_E / sqrt(q) 
        :param delta_phi_m4: orientation of the m=4 profile relative to the position angle of the EPL profile
        :param center_x: center of the profile :param center_y: center of the profile
        :return: gravitational potential in units of theta_E^2."""
```

File: lenstronomy/LensModel/Profiles/epl_multipole_m3m4.py [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+        profile (along 1st axis) :param e2: ellipticity of EPL profile (along 2nd axis)
+        :param a3_a: amplitude of the m=3 multiple deviation from pure elliptical shape
+        related to the physical amplitude of the MULTIPOLE profile by a scaling theta_E
+        / sqrt(q) :param delta_phi_m3: orientation of the m=3 profile relative to the
+        position angle of the EPL profile :param a4_a: amplitude of the m=4 multipole
+        deviation from pure elliptical shape related to the physical amplitude of the
+        MULTIPOLE profile by a scaling theta_E / sqrt(q) :param delta_phi_m4:
+        orientation of the m=4 profile relative to the position angle of the EPL profile
+        :param center_x: center of the profile :param center_y: center of the profile
+        :return: deflection angles in units of theta_E."""
```
Comment (ajshajib) [inline lenstronomy/LensModel/Profiles/epl_multipole_m3m4.py:212]:
```suggestion
        """Compute the derivatives of the potential in units of theta_E 
        :param x: x angular coordinate 
        :param y: y angular coordinate 
        :param theta_E: Einstein radius 
        :param gamma: log-slope of EPL mass profile 
        :param e1: ellipticity of EPL profile (along 1st axis) 
        :param e2: ellipticity of EPL profile (along 2nd axis)
        :param a3_a: amplitude of the m=3 multiple deviation from pure elliptical shape related to the physical amplitude of the MULTIPOLE profile by a scaling theta_E / sqrt(q) 
        :param delta_phi_m3: orientation of the m=3 profile relative to the position angle of the EPL profile 
        :param a4_a: amplitude of the m=4 multipole deviation from pure elliptical shape related to the physical amplitude of the MULTIPOLE profile by a scaling theta_E / sqrt(q) 
        :param delta_phi_m4: orientation of the m=4 profile relative to the position angle of the EPL profile
        :param center_x: center of the profile 
        :param center_y: center of the profile
        :return: deflection angles in units of theta_E."""
```

File: lenstronomy/LensModel/Profiles/epl_multipole_m3m4.py [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+        ellipticity of EPL profile (along 1st axis) :param e2: ellipticity of EPL
+        profile (along 2nd axis) :param a3_a: amplitude of the m=3 multiple deviation
+        from pure elliptical shape related to the physical amplitude of the MULTIPOLE
+        profile by a scaling theta_E / sqrt(q) :param delta_phi_m3: orientation of the
+        m=3 profile relative to the position angle of the EPL profile :param a4_a:
+        amplitude of the m=4 multipole deviation from pure elliptical shape related to
+        the physical amplitude of the MULTIPOLE profile by a scaling theta_E / sqrt(q)
+        :param delta_phi_m4: orientation of the m=4 profile relative to the position
+        angle of the EPL profile :param center_x: center of the profile :param center_y:
+        center of the profile :return: components of the hessian matrix."""
```
Comment (ajshajib) [inline lenstronomy/LensModel/Profiles/epl_multipole_m3m4.py:263]:
```suggestion
        """Compute the second derivatives of the potential that enter the Hessian matrix.
        
        :param x: x angular coordinate 
        :param y: y angular coordinate 
        :param theta_E: Einstein radius 
        :param gamma: log-slope of EPL mass profile 
        :param e1: ellipticity of EPL profile (along 1st axis) 
        :param e2: ellipticity of EPL profile (along 2nd axis) 
        :param a3_a: amplitude of the m=3 multiple deviation from pure elliptical shape related to the physical amplitude of the MULTIPOLE profile by a scaling theta_E / sqrt(q) 
        :param delta_phi_m3: orientation of the m=3 profile relative to the position angle of the EPL profile 
        :param a4_a: amplitude of the m=4 multipole deviation from pure elliptical shape related to the physical amplitude of the MULTIPOLE profile by a scaling theta_E / sqrt(q)
        :param delta_phi_m4: orientation of the m=4 profile relative to the position angle of the EPL profile 
        :param center_x: center of the profile 
        :param center_y: center of the profile 
        :return: components of the hessian matrix."""
```

File: lenstronomy/LensModel/Profiles/epl_multipole_m3m4.py [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+        profile (along 1st axis) :param e2: ellipticity of EPL profile (along 2nd axis)
+        :param a3_a: amplitude of the m=3 multiple deviation from pure elliptical shape
+        related to the physical amplitude of the MULTIPOLE profile by a scaling theta_E
+        / sqrt(q) :param delta_phi_m3: orientation of the m=3 profile relative to the
+        position angle of the EPL profile :param a4_a: amplitude of the m=4 multipole
+        deviation from pure elliptical shape related to the physical amplitude of the
+        MULTIPOLE profile by a scaling theta_E / sqrt(q) :param delta_phi_m4:
+        orientation of the m=4 profile relative to the position angle of the EPL profile
+        :param center_x: center of the profile :param center_y: center of the profile
+        :return: deflection angles in units of theta_E."""
```
Comment (ajshajib) [inline lenstronomy/LensModel/Profiles/epl_multipole_m3m4.py:217]:
```suggestion
        """Compute the derivatives of the potential in units of theta_E 
        
        :param x: x angular coordinate 
        :param y: y angular coordinate 
        :param theta_E: Einstein radius 
        :param gamma: log-slope of EPL mass profile 
        :param e1: ellipticity of EPL profile (along 1st axis) 
        :param e2: ellipticity of EPL profile (along 2nd axis)
        :param a3_a: amplitude of the m=3 multiple deviation from pure elliptical shape related to the physical amplitude of the MULTIPOLE profile by a scaling theta_E / sqrt(q) 
        :param delta_phi_m3: orientation of the m=3 profile relative to the position angle of the EPL profile 
        :param a4_a: amplitude of the m=4 multipole deviation from pure elliptical shape related to the physical amplitude of the MULTIPOLE profile by a scaling theta_E / sqrt(q) 
        :param delta_phi_m4: orientation of the m=4 profile relative to the position angle of the EPL profile
        :param center_x: center of the profile 
        :param center_y: center of the profile
        :return: deflection angles in units of theta_E."""
```

Comment (dangilman) [reply lenstronomy/LensModel/Profiles/epl_multipole_m3m4.py]: @ajshajib amazing! thanks

**Comment by dangilman** [comment]
@sibirrer It appears that the formatting issues have been resolved

File: lenstronomy/LensModel/Solver/solver4point.py [hunk 11]
```diff
@@ -26,6 +27,7 @@ def __init__(self, lensModel, solver_type="PROFILE"):
             "CNFW_ELLIPSE",
             "EPL",
             "EPL_BOXYDISKY",
+            "'EPL_MULTIPOLE_M3M4'",
```
Comment (ajshajib) [inline lenstronomy/LensModel/Solver/solver4point.py:30]:
```suggestion
            "EPL_MULTIPOLE_M3M4",
```

**Comment by dangilman** [comment]
This lens profile was implemented through the decoupled multi-plane formalism pull request, so I'll close this pull request.

**Pull request closed without merging**